src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
-                                                       WHERE fp.id in ( select fv.path_id from fact_values fv where fv.id %s )
-                                                       GROUP BY fp.id
-                                                       HAVING COUNT(fv.id) = 1)" in-list)
-                             fact-ids)
-
-            (sql/do-prepared (format "DELETE FROM fact_values WHERE id in (
-                                          SELECT fact_value_id FROM facts where factset_id = ? and fact_value_id %s
-                                          EXCEPT
-                                          SELECT fact_value_id FROM facts where factset_id <> ? and fact_value_id %s)" in-list in-list)
-                             (concat (cons factset fact-ids)
-                                     (cons factset fact-ids))))
-
-          (sql/delete-rows :facts
-                           (into [(str "factset_id=? and fact_value_id "
-                                       in-list) factset]
-                                 fact-ids))
-
-          ;; HSQLDB doesn't support delayed constraints. This query
-          ;; achieves the same goals as the DELETE FROM fact_values
-          ;; above, but does so in a less efficient manner. Having
-          ;; these two separated allows Postgres queries to execute
-          ;; faster, but still support HSQLDB
-          (when-not (sutils/postgres?)
-            (sql/do-prepared (format "DELETE FROM fact_values WHERE id in (
-                                        SELECT fv.id
-                                        FROM fact_values fv left join facts f on fv.id = f.fact_value_id
-                                        WHERE f.fact_value_id is NULL and fv.id %s )" in-list)
-                             fact-ids)
-            (sql/do-prepared (format "DELETE FROM fact_paths fp
-                                      WHERE fp.id in ( SELECT fp.id
-                                                       FROM fact_paths fp left join fact_values fv on fp.id = fv.path_id
-                                                       WHERE fv.path_id is NULL )"))))))))
-
-(pls/defn-validated update-facts!
+  (sql/transaction
+   (sql/insert-record :factsets
+                      {:certname name
+                       :timestamp (to-timestamp timestamp)
+                       :environment_id (ensure-environment environment)
+                       :producer_timestamp (to-timestamp producer-timestamp)})
+   ;; Ensure that all the required paths and values exist, and then
+   ;; insert the new facts.
+   (let [paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         path-ids (map (realize-paths! pathstrs) pathstrs)
+         vhash-ids (map :id (map (realize-values! valuemaps) vhashes))
```

Comment:
I think we should be able to just use something like
```
(map :id (realize-values! valuemaps))
```
Then we won't need vhashes and it won't traverse valuemaps again